pkg/controller/endpointslice/topologycache/topologycache.go

```diff
@@ -84,12 +88,13 @@ func (t *TopologyCache) GetOverloadedServices() []string {
 
 // AddHints adds or updates topology hints on EndpointSlices and returns updated
 // lists of EndpointSlices to create and update.
-func (t *TopologyCache) AddHints(si *SliceInfo) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice) {
+func (t *TopologyCache) AddHints(si *SliceInfo, service *corev1.Service, recorder record.EventRecorder) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice) {
 	totalEndpoints := si.getTotalReadyEndpoints()
 	allocations := t.getAllocations(totalEndpoints)
 
 	if allocations == nil {
-		klog.V(2).InfoS("Insufficient endpoints, removing hints from service", "serviceKey", si.ServiceKey)
+		klog.V(2).InfoS("Insufficient endpoints, removing hints from service", "serviceKey", si.ServiceKey, "addressType", si.AddressType)
+		recorder.Eventf(service, corev1.EventTypeWarning, "TopologyAwareHintsDisabled", "Insufficient number of endpoints, removing %s hints from service %q", si.AddressType, si.ServiceKey)
```

Comment:
I still think I prefer returning a list of events to publish here, since TopologyCache generally does not write anything, it just returns a list of objects to write.